### PR TITLE
Check for output overflows in the non-streamable implementation

### PR DESCRIPTION
**Pull request by clecat**
These few commits adds checks on the size of the output everywhere when it was forgotten.
This problem was revealed by [this issue](https://github.com/mirage/decompress/issues/126) on `ocaml-git`. Be aware that this PR does not solve the problem as it comes from `ocaml-git` itself, but it will allow for a more clean error handling in the future.
It also adds few test cases for the inflation.

It also contains some cleaning linked to the output writing:
- removed the function `_fill` function which was not used anymore.  (line 1528 to 1546)
- removed an artifact from the streamable api where we would not write all the output if not possible. (line 1618 to 1623)
- removed strange behavior in `add_bits` and `flush_bits` with `o_pos` incrementation. (line 3240 and 3248)
- rewrite `memcpy` function so it uses `unsafe_set_uint32` whenever possible (line 3266 to 3273)

### CODE REVIEW

File: lib/de.ml [hunk 1]
```diff
@@ -3263,24 +3248,28 @@ module Def = struct
...
+        let i = i * 4 in
+        let v = unsafe_get_uint32 src (src_off + i) in
+        unsafe_set_uint32 dst (dst_off + i) v
       done
-      ; os.i_pos <- os.i_pos + !i
+      ; for i = 0 to len0 - 1 do
+          let i = (len1 * 4) + i in
+          let v = unsafe_get_uint8 src (src_off + i) in
+          unsafe_set_uint8 dst (dst_off + i) v
+        done
```
Comment (dinosaure) [inline lib/de.ml:3263]: Can you extract this part to another PR? Just to be sure to separate the PR which fixes the bug and a PR which improve performances.

**Comment by dinosaure** [comment]
Seems good, I will try to fix the error about `ppx_deriving_yojson` and I will rebase your PR with that. Thanks :+1:.

File: lib/de.ml (hunk 1 above)
Comment (clecat) [reply lib/de.ml:3263]: Of course
